backend/layers/processing/rollback.py

```diff
@@ -0,0 +1,301 @@
+        the private collection to its pre-migration state.
+        """
+        collection_version = self.business_logic.get_collection_version(collection_version_id)
+        if collection_version is None:
+            raise CollectionNotFoundException(f"CollectionVersion {collection_version_id} not found")
+        for dataset in collection_version.datasets:
+            try:
+                self.rollback_private_dataset(dataset.version_id, collection_version_id)
+            except Exception as e:
+                logger.info(f"Failed to rollback DatasetVersion {dataset.version_id}, {e}")
+                continue
+        return collection_version.datasets
+
+    def rollback_private_dataset_list(self) -> None:
+        """
+        Rolls back the Datasets in the CollectionVersions associated with each DatasetVersionId passed in, to their
+        respective previous, most recently created DatasetVersion. Then, triggers deletion of the DB references and S3
+        assets for the rolled back DatasetVersions.
+        """
+        rolled_back_datasets = []
+        for dataset_version_id in self.entity_id_list:
+            dataset_version_id = DatasetVersionId(dataset_version_id)
+            try:
+                rolled_back_dataset = self.rollback_private_dataset(dataset_version_id)
+            except Exception as e:
+                logger.info(f"Failed to rollback DatasetVersion {dataset_version_id}, {e}")
+                continue
+            rolled_back_datasets.append(rolled_back_dataset)
+        self._clean_up_rolled_back_datasets(rolled_back_datasets)
+
+    def rollback_private_dataset(
+        self, dataset_version_id: DatasetVersionId, collection_version_id: CollectionVersionId = None
+    ) -> DatasetVersion:
+        """
+        For a given DatasetVersionId and unpublished CollectionVersionId, rolls back the associated Dataset in the
+        CollectionVersion to its previous, most recently created DatasetVersion and deletes the given DatasetVersion.
+
+        :param dataset_version_id: DatasetVersionId of the DatasetVersion to rollback
+        :param collection_version_id: CollectionVersionId of the CollectionVersion to rollback the DatasetVersion
+        in. If not passed in, the CollectionVersionId will be determined from the DatasetVersionId.
+        :return: DatasetVersion that was rolled back
+        """
+        cv_id = collection_version_id
+        dataset_version = self.business_logic.get_dataset_version(dataset_version_id)
+        if cv_id is None:
+            # account for collection potentially having a migration revision and a non-migration revision
+            collection_versions = self.business_logic.get_unpublished_collection_versions_from_canonical(
+                dataset_version.collection_id
+            )
+            for cv in collection_versions:
+                if cv_id is not None:
```

Comment:
Is this condition ever met? It looks like the `break` on line 200 should make this conditional superfluous.

Adding a comment would help clarify.

Reply:
I believe breaking out a nested for loop does not break you out of the outer for loop